File: go/hamming/hamming.go
// Package hamming implements functions to calculate the Hamming distance between DNA strands.
package hamming

import "errors"

// Distance calculates the Hamming distance between two DNA strands.
func Distance(a, b string) (int, error) {
	if len(a) != len(b) {
		return -1, errors.New("the two DNA strands should have the same length")
	}

	distance := 0
	for i := 0; i < len(a); i++ {
		if a[i] != b[i] {
			distance++
		}
	}
	return distance, nil
}

File: typescript/hamming/hamming.ts
export default class Hamming {
  compute(strandA: string, strandB: string) {
    if (strandA.length !== strandB.length)
      throw new Error("DNA strands must be of equal length.");

    let distance = 0;
    for (let i = 0; i < strandA.length; i++)
      if (strandA[i] != strandB[i]) distance++;
    return distance;
  }
}
